crates/typst-pdf/src/embed.rs

```diff
@@ -0,0 +1,90 @@
+use crate::catalog::pdf_date;
+use crate::{PdfAConformanceLevel, PdfChunk, WithGlobalRefs};
+use ecow::EcoString;
+use pdf_writer::{Finish, Name, Ref, Str, TextStr};
+use std::collections::HashMap;
+use typst_library::diag::{bail, SourceResult};
+use typst_library::foundations::NativeElement;
+use typst_library::pdf::embed::{Embed, EmbedElem, EmbeddedFileRelationship};
+use typst_syntax::Span;
+
+pub fn write_embedded_files(
+    ctx: &WithGlobalRefs,
+) -> SourceResult<(PdfChunk, HashMap<EcoString, Ref>)> {
+    let mut chunk = PdfChunk::new();
+
+    let elements = ctx.document.introspector.query(&EmbedElem::elem().select());
+    if !elements.is_empty()
+        && Some(PdfAConformanceLevel::A_2) == ctx.options.standards.pdfa
+    {
+        bail!(
+            Span::detached(),
+            "file embeddings are currently only supported for PDF/A-3"
+        );
+    }
+
+    let mut embedded_files = HashMap::default();
+    for elem in elements.iter() {
+        let packed_elem = elem.to_packed::<EmbedElem>().unwrap();
+        let embed = Embed::from_element(packed_elem);
+        embedded_files.insert(embed.name().clone(), embed_file(ctx, &mut chunk, &embed));
+    }
+
+    Ok((chunk, embedded_files))
+}
+
+fn embed_file(ctx: &WithGlobalRefs, chunk: &mut PdfChunk, embed: &Embed) -> Ref {
+    let embedded_file_stream_ref = chunk.alloc.bump();
+    let file_spec_dict_ref = chunk.alloc.bump();
+
+    let length = embed.data().len();
+
+    let mut embedded_file =
+        chunk.embedded_file(embedded_file_stream_ref, embed.data().as_ref());
+    embedded_file.pair(Name(b"Length"), length as i32);
+    if let Some(mime_type) = embed.mime_type() {
+        embedded_file.subtype(Name(mime_type.as_bytes()));
+    }
+    let date = ctx
+        .document
+        .info
+        .date
+        .unwrap_or(ctx.options.timestamp)
+        .and_then(|date| pdf_date(date, ctx.document.info.date.is_auto()));
+    if let Some(pdf_date) = date {
+        embedded_file.params().modification_date(pdf_date).finish();
```

Comment:
Does it make sense to say that the modification date of the _embedded file_ is the _document's_ date?

Reply:
It is at least the latest time that it could have been modified at. Alternatively, we could add another optional parameter, but I though this is a good solution for the beginning.

Reply:
If we want to support this, I would suggest making the file creation and modification date available via the world as an optional parameter and then do `modification-date` with a default value of `auto` an additional named parameter. As with `document`'s `date`, this can also be `none` or an explicit date. If a path is passed, the modification date from the world, if any, should be used, if it's a bytes / string value, it should be the world's `today`. This way, we provide the actual expected value for files present on the disk, allow opting out and overriding, and provide a way for world implementors to opt out of this feature if their FS implementation does not track modification dates.

Reply:
> I would suggest making the file creation and modification date available via the world as an optional parameter

I am unsure how you imagined the "as an optional parameter" part. I tested adding another method to the `World` trait to get the optional modification date for a file by `FileId` (d4cf66c93686ab5dc775803daa47c005b9f685f9). Did you have something else in mind?

Reply:
@reknih The extension of the world trait is pushed and it seems to work fine with the CLI world. However, I encountered issues adding the optional smart parameter to the element. As far as I understood, I should handle the smart parameter in `parse`. The following is my current attempt at that:

```rust
    #[parse(
        let last_modified = if let Some(Spanned { v: smart, span: _ }) =
            args.named::<Spanned<Smart<Option<Datetime>>>>("last-modified")?
        {
            Some(Smart::Custom(match smart {
                Smart::Auto => engine.world.last_modified(id).map_err(|file_error| {
                    eco_vec![error!(
                        path_span,
                        "failed to read metadata for file `{path}`: {file_error:?}";
                        hint: "todo",
                    )]
                })?,
                Smart::Custom(date) => date,
            }))
        } else {
            Some(Smart::Custom(engine.world.last_modified(id).map_err(|file_error| {
                eco_vec![error!(
                    path_span,
                    "failed to read metadata for file `{path}`: {file_error:?}";
                    hint: "todo",
                )]
            })?))
        };
        last_modified
    )]
    pub last_modified: Smart<Option<Datetime>>,
```

But this does not compile, because looking up the modification date through the world requires the `FileId`, which I don't have at both positions where this code gets inserted. These are the relevant parts of expanding the macros:

```rust
    impl ::typst_library::foundations::Construct for EmbedElem {
        fn construct(engine: &mut ::typst_library::engine::Engine, args: &mut ::typst_library::foundations::Args ) -> ::typst_library::diag::SourceResult<::typst_library::foundations::Content> {
            let Spanned { v: path, span: path_span } =
                args.expect::<Spanned<EcoString>>("path to the file to be embedded")?;
            ;
            let id = path_span.resolve_path(&path).at(path_span)?;
            ;
            let data = engine.world.file(id).at(path_span)?;
            let path = path
                ;
            let data = Readable::Bytes(data);
            let name = args.named("name")?;
            let description = args.named("description")?;
            let mime_type = args.named("mime-type")?;
            let relationship = args.named("relationship")?;
            let last_modified = if let Some(Spanned { v: smart, span: _ }) =
                args.named::<Spanned<Smart<Option<Datetime>>>>("last-modified")?
            {
                Some(Smart::Custom(match smart {
                    Smart::Auto => engine.world.last_modified(id).map_err(|file_error| {
                        eco_vec![error!(
                        path_span,
                        "failed to read metadata for file `{path}`: {file_error:?}";
                        hint: "todo",
                    )]
                    })?,
                    Smart::Custom(date) => date,
                }))
            } else {
                Some(Smart::Custom(engine.world.last_modified(id).map_err(|file_error| {
                    eco_vec![error!(
                    path_span,
                    "failed to read metadata for file `{path}`: {file_error:?}";
                    hint: "todo",
                )]
                })?))
            };
            let last_modified = last_modified
                ;
            Ok(::typst_library::foundations::Content::new(Self { path, data, name, description, mime_type, relationship, last_modified }))
        }
    }


    impl ::typst_library::foundations::Set for EmbedElem {
        fn set(engine: &mut ::typst_library::engine::Engine, args: &mut ::typst_library::foundations::Args ) -> ::typst_library::diag::SourceResult<::typst_library::foundations::Styles> {
            let mut styles = ::typst_library::foundations::Styles::new();
            if let Some(value) = args.named("name")? { styles.set(Self::set_name(value)); }
            if let Some(value) = args.named("description")? { styles.set(Self::set_description(value)); }
            if let Some(value) = args.named("mime-type")? { styles.set(Self::set_mime_type(value)); }
            if let Some(value) = args.named("relationship")? { styles.set(Self::set_relationship(value)); }
            let last_modified = if let Some(Spanned { v: smart, span: _ }) =
                args.named::<Spanned<Smart<Option<Datetime>>>>("last-modified")?
            {
                Some(Smart::Custom(match smart {
                    Smart::Auto => engine.world.last_modified(id).map_err(|file_error| {
                        eco_vec![error!(
                        path_span,
                        "failed to read metadata for file `{path}`: {file_error:?}";
                        hint: "todo",
                    )]
                    })?,
                    Smart::Custom(date) => date,
                }))
            } else {
                Some(Smart::Custom(engine.world.last_modified(id).map_err(|file_error| {
                    eco_vec![error!(
                    path_span,
                    "failed to read metadata for file `{path}`: {file_error:?}";
                    hint: "todo",
                )]
                })?))
            };
            if let Some(value) = last_modified
            { styles.set(Self::set_last_modified(value)); }
            Ok(styles)
        }
    }
```

The `Construct` implementation seems fine because I have access to the path parameter, but `Set` does not compile because there is no path or file ID.
My understanding of Typst's macro machinery is quite limited. Is there any extended documentation on this that I missed? Or, if not, could you explain it to me?

Reply:
To be honest, I'm not in favor of this change. The value of having the real file modification in the PDF seems quite small to me, while it hurts reproducibility and increases the complexity of the `World` interface. Writing the document's date itself was already a point of contention with some users that put a higher priority on reproducibility. This goes way further. In particular, the modification date might get lost through copying, downloading, sharing, uploading, etc. I'd like to add as little data into the mix that isn't obviously part of the inputs from a user PoV.

In principle, I understand the desire to follow the standard as much as possible. But I feel like these standards were developed primarily through the desktop-publishing lens where files are edited in a visual editor and exported rather than from the lens of a command line processor that wants to generate reproducible output. There a little bit of metadata might not hurt, but I feel like our case is different.

Reply:
I reverted the World extension and use the document date if there is one. I think the first version of this should add some useful date to have valid embeddings and allow the creation of valid ZUGFeRD documents.

If there is consent on how to add a more accurate last modified date, we could do it in a follow-up, in my opinion.